src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,510 @@
+	Scheme SignatureScheme
+
+	// The credential's delegation.
+	Signature []byte
+}
+
+// marshalSubjectPublicKeyInfo returns a DER encoded SubjectPublicKeyInfo structure
+// (as defined in the X.509 standard) for the credential.
+// TODO: maybe we can move this as well to x509
+// TODO: add the other signatures flavors, as defined in common.go
+func (cred *credential) marshalPublicKeyInfo() ([]byte, error) {
+	switch cred.scheme {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512:
+		serializedPublicKey, err := x509.MarshalPKIXPublicKey(cred.publicKey)
+		if err != nil {
+			return nil, err
+		}
+		return serializedPublicKey, nil
+
+	default:
+		return nil, fmt.Errorf("unsupported signature scheme: 0x%04x", cred.scheme)
+	}
+}
+
+// unmarshalPublicKeyInfo parses a DER encoded PublicKeyInfo
+// structure into a public key and its corresponding algorithm.
+// TODO: add the other signatures flavors, as defined in common.go
+func unmarshalPublicKeyInfo(serialized []byte) (crypto.PublicKey, SignatureScheme, error) {
+	publicKey, err := x509.ParsePKIXPublicKey(serialized)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	switch pk := publicKey.(type) {
+	case *ecdsa.PublicKey:
+		curveName := pk.Curve.Params().Name
+		if curveName == "P-256" {
+			return pk, ECDSAWithP256AndSHA256, nil
+		} else if curveName == "P-384" {
+			return pk, ECDSAWithP384AndSHA384, nil
+		} else if curveName == "P-521" {
+			return pk, ECDSAWithP521AndSHA512, nil
+		} else {
+			return nil, 0, fmt.Errorf("curve %s s not supported", curveName)
+		}
+
+	default:
+		return nil, 0, fmt.Errorf("unsupported delgation key type: %T", pk)
+	}
+}
+
+// marshal encodes the credential.
+// TODO: might be better if this is a function and not a method
+func (cred *credential) marshal() ([]byte, error) {
+	// Write the valid_time field.
+	ser := make([]byte, 6)
+	binary.BigEndian.PutUint32(ser, uint32(cred.validTime/time.Second))
+
+	// Encode the public key and assert that the encoding is no longer than 2^16
+	// bytes (per the spec).
+	// TODO: double check this
+	serPubKey, err := cred.marshalPublicKeyInfo()
+	if err != nil {
+		return nil, err
+	}
+	if len(serPubKey) > dcMaxPublicKeyLen {
+		// TODO: change this error
+		return nil, errors.New("public key length is not valid")
+	}
+
+	binary.BigEndian.PutUint16(ser[4:], uint16(len(serPubKey)))
+	return append(ser, serPubKey...), nil
+}
+
+// unmarshalCredential decodes serialized bytes and returns a credential if possible.
+func unmarshalCredential(ser []byte) (*credential, error) {
+	// Bytes 0-3 are the validity time field; bytes 4-6 are the length of the
+	// serialized PublicKeyInfo.
+	if len(ser) < 6 {
+		// TODO: change this error
+		return nil, errors.New("credential is invalid")
+	}
+
+	// Parse the validity time.
+	validTime := time.Duration(binary.BigEndian.Uint32(ser)) * time.Second
+
+	// Parse the PublicKeyInfo.
+	pk, scheme, err := unmarshalPublicKeyInfo(ser[6:])
```

Comment:
the length of the public key  must be used to bound the slice passed to `unmarshalPublicKeyInfo`. Something like this ->

```suggestion
	pkSize := binary.BigEndian.Uint16(ser[4:6])
	pk, scheme, err := unmarshalPublicKeyInfo(ser[6:pkSize])
```